File: android/src/main/java/com/github/wuxudong/rncharts/data/BarDataExtract.java
package com.github.wuxudong.rncharts.data;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableType;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.data.ChartData;
import com.github.mikephil.charting.interfaces.datasets.IDataSet;
import com.github.wuxudong.rncharts.utils.BridgeUtils;
import com.github.wuxudong.rncharts.utils.ChartDataSetConfigUtils;

import java.util.ArrayList;

/**
 * Created by xudong on 02/03/2017.
 */

public class BarDataExtract extends DataExtract<BarData, BarEntry> {
    @Override
    BarData createData() {
        return new BarData();
    }

    @Override
    IDataSet<BarEntry> createDataSet(ArrayList<BarEntry> entries, String label) {
        return new BarDataSet(entries, label);
    }

    @Override
    BarEntry createEntry(ReadableArray values, int index) {
        BarEntry entry;

        float x = index;
        if (ReadableType.Map.equals(values.getType(index))) {
            ReadableMap map = values.getMap(index);

            if (map.hasKey("x")) {
                x = (float) map.getDouble("x");
            }

            if (ReadableType.Array.equals(map.getType("y"))) {
                entry = new BarEntry(x, BridgeUtils.convertToFloatArray(map.getArray("y")));
            } else if (ReadableType.Number.equals(map.getType("y"))) {
                entry = new BarEntry(x, (float) map.getDouble("y"));
            } else {
                throw new IllegalArgumentException("Unexpected entry type: " + values.getType(index));
            }

            entry.setData(map);

        } else if (ReadableType.Array.equals(values.getType(index))) {
            entry = new BarEntry(x, BridgeUtils.convertToFloatArray(values.getArray(index)));
        } else if (ReadableType.Number.equals(values.getType(index))) {
            entry = new BarEntry(x, (float) values.getDouble(index));
        } else {
            throw new IllegalArgumentException("Unexpected entry type: " + values.getType(index));
        }

        return entry;
    }

    @Override
    void dataSetConfig(IDataSet<BarEntry> dataSet, ReadableMap config) {
        BarDataSet barDataSet = (BarDataSet) dataSet;

        ChartDataSetConfigUtils.commonConfig(barDataSet, config);
        ChartDataSetConfigUtils.commonBarLineScatterCandleBubbleConfig(barDataSet, config);

        if (BridgeUtils.validate(config, ReadableType.Number, "barShadowColor")) {
            barDataSet.setBarShadowColor(config.getInt("barShadowColor"));
        }
        if (BridgeUtils.validate(config, ReadableType.Number, "highlightAlpha")) {
            barDataSet.setHighLightAlpha(config.getInt("highlightAlpha"));
        }
        if (BridgeUtils.validate(config, ReadableType.Array, "stackLabels")) {
            barDataSet.setStackLabels(BridgeUtils.convertToStringArray(config.getArray("stackLabels")));
        }
    }

    @Override
    void dataConfig(BarData data, ReadableMap config) {
        super.dataConfig(data, config);

        if (BridgeUtils.validate(config, ReadableType.Number, "barWidth")) {
            data.setBarWidth((float) config.getDouble("barWidth"));
        }

        if (BridgeUtils.validate(config, ReadableType.Number, "group")) {
            ReadableMap propMap = config.getMap("group");
            if (BridgeUtils.validate(propMap, ReadableType.Number, "fromX") &&
                    BridgeUtils.validate(propMap, ReadableType.Number, "groupSpace") &&
                    BridgeUtils.validate(propMap, ReadableType.Number, "barSpace")
                    ) {
                float fromX = (float) propMap.getDouble("fromX");
                float groupSpace = (float) propMap.getDouble("groupSpace");
                float barSpace = (float) propMap.getDouble("barSpace");

                data.groupBars(fromX, groupSpace, barSpace);
            }
        }


    }
}

File: ios/ReactNativeCharts/bar/BarDataExtract.swift
//  Created by xudong wu on 02/03/2017.
//  Copyright © 2017 wuxudong. All rights reserved.
//

import Foundation

import SwiftyJSON
import Charts

class BarDataExtract : DataExtract {
    override open func createData() -> ChartData {
        return BarChartData();
    }
    
    override open func dataConfig(_ data: ChartData, config: JSON) {
    
        let barData = data as! BarChartData
      
        if config["barWidth"].double != nil {
            barData.barWidth = config["barWidth"].doubleValue
        }
        
        if config["group"].exists() {
            let fromX = config["group"]["fromX"].doubleValue;
            let groupSpace = config["group"]["groupSpace"].doubleValue;
            let barSpace = config["group"]["barSpace"].doubleValue;
            
            barData.groupBars(fromX: fromX, groupSpace: groupSpace, barSpace: barSpace)
        }
        
       
    }
    
    override open func createDataSet(_ entries: [ChartDataEntry]?, label: String?) -> IChartDataSet {
        return BarChartDataSet(values: entries, label: label)
    }
    
    override open func dataSetConfig(_ dataSet: IChartDataSet, config: JSON) {
        let barDataSet = dataSet as! BarChartDataSet
        
        ChartDataSetConfigUtils.commonConfig(barDataSet, config: config);
        ChartDataSetConfigUtils.commonBarLineScatterCandleBubbleConfig(barDataSet, config: config);
        
        if config["barShadowColor"].int != nil {
            barDataSet.barShadowColor = RCTConvert.uiColor(config["barShadowColor"].intValue)
        }
        
        if config["highlightAlpha"].number != nil {
            barDataSet.highlightAlpha = BridgeUtils.toIOSAlpha(config["highlightAlpha"].numberValue)
        }
        
        if config["stackLabels"].array != nil {
            barDataSet.stackLabels = config["stackLabels"].arrayValue.map({ return $0.stringValue })
        }
    }
    
    override func createEntry(_ values: [JSON], index: Int) -> BarChartDataEntry {
        var entry: BarChartDataEntry;
        
        var x = Double(index);
        let value = values[index];
        
        if value.dictionary != nil {
            let dict = value;
            
            if dict["x"].double != nil {
                x = Double(dict["x"].doubleValue);
            }
            
            if dict["y"].array != nil {
                entry = BarChartDataEntry(x: x, yValues: (dict["y"].arrayValue.map({ y in y.doubleValue })));
            } else if dict["y"].number != nil {
                entry = BarChartDataEntry(x: x, y: dict["y"].doubleValue);
            } else {
                fatalError("invalid data " + values.description);
            }
            
            entry.data = dict as AnyObject?;
        } else if value.array != nil {
            entry = BarChartDataEntry(x: x, yValues: (value.arrayValue.map({ y in y.doubleValue })));
        } else if value.double != nil {
            entry = BarChartDataEntry(x: x, y: value.doubleValue);
        } else {
            fatalError("invalid data " + values.description);
        }
        
        return entry;
    }
}
